manticore/core/smtlib/expression.py

```diff
@@ -966,279 +1227,249 @@ def index(self):
     def value(self):
         return self.operands[2]
 
-    def __getstate__(self):
-        state = {}
-        array = self
-        items = []
-        while isinstance(array, ArrayStore):
-            items.append((array.index, array.value))
+    def select(self, index):
+
+        """Gets an element from the Array.
+        If the element was not previously the default is used.
+        """
+        index = local_simplify(self.cast_index(index))
+
+        # Emulate list[-1]
+        has_length = self.length is not None
+        if has_length:
+            index = local_simplify(BitvecITE(index < 0, self.length + index, index))
+
+        if isinstance(index, Constant):
+            if has_length and index.value >= self.length:
+                raise IndexError
+            if index.value in self.concrete_cache:
+                return self.concrete_cache[index.value]
+
+        default = self.default
+        if default is None:
+            # No default. Returns normal array select
+            return ArraySelect(self, index)
+
+        # if a default is defined we need to check if the index was previously written
+        return BitvecITE(self.is_known(index), ArraySelect(self, index), self.cast_value(default))
+
+        # build a big ITE expression
+        array, offset, items = self, 0, []
+        while not isinstance(array, ArrayVariable):
```

Comment:
The `return` statement creates dead code. I can't figure out what the correct thing to do is, though.

Is everything after the `return` supposed to be deleted?

Reply:
Yes. But I'm not super comfortable with what happens before the return.

Reply:
Oh so the thing after the return tries not to use Arrays. An build all the array logic as a big ITE.